### PR TITLE
Added wolfSSL compatability for Asio C++ library

### PR DESCRIPTION
**Pull request by MJSPollard**

### CODE REVIEW

**Comment by JacobBarthelmeh** [comment]
test this please Jenkins

**Comment by MJSPollard** [comment]
retest this please jenkins

File: src/ssl.c [hunk 1]
```diff
@@ -30936,6 +30985,230 @@ int wolfSSL_get_state(const WOLFSSL* ssl)
...
+        WOLFSSL_MSG("one or more inputs were NULL");
+        return BAD_FUNC_ARG;
+    }
+
+    if((bio = BIO_new(BIO_s_mem())) == NULL){
+        WOLFSSL_MSG("Set bio Failed");
+        return WOLFSSL_FAILURE;
+    }
+
+    pkey = wolfSSL_PKEY_new_ex(bio->heap);
```
[the remark this answers is not in the document]
Comment (JacobBarthelmeh) [reply src/ssl.c]: I think the structure pkey will not be needed though if using the function wolfSSL_CTX_use_PrivateKey_buffer after changes

File: src/ssl.c [hunk 2]
```diff
@@ -30936,6 +30985,230 @@ int wolfSSL_get_state(const WOLFSSL* ssl)
...
+
+    wolfSSL_BIO_free(bio);
+    wolfSSL_EVP_PKEY_free(pkey);
+    XFREE((unsigned char*)derBuf, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+    return ret;
+}
+#endif /* NO_RSA && !HAVE_FAST_RSA */
+
+
+WOLFSSL_EVP_PKEY* wolfSSL_d2i_PrivateKey_bio(WOLFSSL_BIO* bio, WOLFSSL_EVP_PKEY** out)
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: comments on return value of function

File: src/ssl.c [hunk 3]
```diff
@@ -27773,7 +27753,76 @@ WOLFSSL_RSA *wolfSSL_d2i_RSAPublicKey(WOLFSSL_RSA **r, const unsigned char **pp,
...
+        return NULL;
+    }
+    if(r != NULL)
+        *r = rsa;
+    return rsa;
+}
+
 #if !defined(HAVE_FAST_RSA)
+#if defined(WOLFSSL_KEY_GEN)
+int wolfSSL_i2d_RSAPrivateKey(WOLFSSL_RSA *rsa, const unsigned char **pp)
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: comment on function and return value, also check compatibility with pointer value before and after the operation. Should pp not be a const type?

File: src/ssl.c [hunk 4]
```diff
@@ -27773,7 +27753,76 @@ WOLFSSL_RSA *wolfSSL_d2i_RSAPublicKey(WOLFSSL_RSA **r, const unsigned char **pp,
...
+        WOLFSSL_MSG("SetRsaInternal Failed");
+        return ret;
+    }
+
+    /* 5 > size of n, d, p, q, d%(p-1), d(q-1), 1/q%p, e + ASN.1 additional
+     *  informations
+     */
+    derMax = 5 * wolfSSL_RSA_size(rsa) + AES_BLOCK_SIZE;
+
+    der = (byte*)XMALLOC(derMax, NULL, DYNAMIC_TYPE_TMP_BUFFER);
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: OpenSSL uses the passed in buffer to hold the DER value

File: src/ssl.c [hunk 5]
```diff
@@ -30936,6 +30985,230 @@ int wolfSSL_get_state(const WOLFSSL* ssl)
...
+            XFREE((unsigned char*)memptr, bio->heap, DYNAMIC_TYPE_TMP_BUFFER);
+            return NULL;
+        }
+
+        key = wolfSSL_d2i_PrivateKey(type, &key, &mem, memSz);
+        if (out != NULL && key != NULL) {
+            *out = key;
+        }
+    }
+    XFREE((unsigned char*)memptr, bio->heap, DYNAMIC_TYPE_TMP_BUFFER);
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: at the end of the function we need to take into account data at  the end of the WOLFSSL_BIO buffer that may not be related to DER

Comment (MJSPollard) [reply src/ssl.c]: I'm little stuck on this one, any ideas?

File: src/ssl.c [hunk 6]
```diff
@@ -31067,6 +31154,380 @@ int wolfSSL_get_state(const WOLFSSL* ssl)
...
+    }
+
+    bufPtr = maxKeyBuf;
+    if (wolfSSL_BIO_read(bio, (unsigned char*)bioMem, (int)bioMemSz) == bioMemSz) {
+        if((key = wolfSSL_d2i_RSAPrivateKey(NULL, &bioMem, bioMemSz)) == NULL) {
+            XFREE((unsigned char*)bioMem, bio->heap, DYNAMIC_TYPE_TMP_BUFFER);
+            return NULL;
+        }
+
+        //This function is used to get the total length of the rsa key.
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: please use /**/ style comments instead of // comments

File: src/ssl.c [hunk 7]
```diff
@@ -31067,6 +31154,380 @@ int wolfSSL_get_state(const WOLFSSL* ssl)
...
+                return NULL;
+            }
+
+            for(i = derLength; i < bioMemSz; i++) {
+                *(extraBioMem + j) = *(bioMem + i);
+                j++;
+            }
+
+            BIO_write(bio, extraBioMem, extraBioMemSz);
+            if(wolfSSL_BIO_pending(bio) <= 0) {
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: check if and for statements for coding standard. Use "if (" instead of "if(".

File: src/ssl.c [hunk 8]
```diff
@@ -31067,6 +31154,380 @@ int wolfSSL_get_state(const WOLFSSL* ssl)
...
+
+    if (wolfSSL_BIO_read(bio, (unsigned char*)mem, memSz) == memSz) {
+        /* Determines key type and returns the new private EVP_PKEY object */
+        if((key = wolfSSL_d2i_PrivateKey_EVP(NULL, &mem, (long)memSz)) == NULL) {
+            WOLFSSL_MSG("d2i_PrivateKey_EVP() failure");
+            XFREE(mem, bio->heap, DYNAMIC_TYPE_TMP_BUFFER);
+            return NULL;
+        }
+
+        //Write extra data back into bio object if necessary.
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: change the comment type here

File: src/ssl.c [hunk 9]
```diff
@@ -31067,6 +31154,380 @@ int wolfSSL_get_state(const WOLFSSL* ssl)
...
+                XFREE(mem, bio->heap, DYNAMIC_TYPE_TMP_BUFFER);
+                return NULL;
+            }
+
+            for(i = derLength; i < memSz; i++) {
+                *(extraBioMem + j) = *(mem + i);
+                j++;
+            }
+
+            BIO_write(bio, extraBioMem, extraBioMemSz);
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: use wolfSSL versions of the API when called from inside wolfSSL code i.e. src/ssl.c src/internal.c

File: src/ssl.c [hunk 10]
```diff
@@ -31067,6 +31154,380 @@ int wolfSSL_get_state(const WOLFSSL* ssl)
...
+                XFREE((unsigned char*)bioMem, bio->heap, DYNAMIC_TYPE_TMP_BUFFER);
+                return NULL;
+            }
+
+            for(i = derLength; i < bioMemSz; i++) {
+                *(extraBioMem + j) = *(bioMem + i);
+                j++;
+            }
+
+            BIO_write(bio, extraBioMem, extraBioMemSz);
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: use wolfSSL versions of the API when called from inside wolfSSL code i.e. src/ssl.c src/internal.c

File: src/ssl.c [hunk 11]
```diff
@@ -27923,6 +27955,61 @@ WOLFSSL_RSA *wolfSSL_d2i_RSAPublicKey(WOLFSSL_RSA **r, const unsigned char **pp,
...
+
+    /* RSA key to DER */
+    if((ret = wc_RsaKeyToDer((RsaKey *)rsa->internal, der, derMax)) < 0) {
+        WOLFSSL_MSG("RsaKeyToDer failed");
+        XFREE(der, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+        der = NULL;
+        return ret;
+    }
+
+    //ret is the size of the der buffer
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: comment style

File: src/ssl.c [hunk 12]
```diff
@@ -31067,6 +31154,380 @@ int wolfSSL_get_state(const WOLFSSL* ssl)
...
+    key = maxDerBuf;
+    /* convert RSA struct to der encoded buffer and get the size */
+    if((derSize = wolfSSL_i2d_RSAPrivateKey(rsa, &key)) <= 0) {
+        WOLFSSL_MSG("wolfSSL_i2d_RSAPrivateKey failed");
+        return WOLFSSL_FAILURE;
+    }
+    ret = wolfSSL_CTX_use_PrivateKey_buffer(ctx, (const unsigned char*)maxDerBuf, 
+                                                    derSize, SSL_FILETYPE_ASN1);
+    if(ret != WOLFSSL_SUCCESS) {
+        WOLFSSL_MSG("CTX_USE_PrivateKey_buffer() failure");
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: print wolfSSL version of the function in debug message here

**Comment by JacobBarthelmeh** [comment]
Thanks!

**Pull request merged**